File: src/d4.py
#%%
with open("../data/d4.txt", "r") as f:
    p1_total = 0
    p2_total = 0
    for pair in f.readlines():
        # ---- Part 1 ----
        e1 = pair.split(",")[0].split("-")
        e2 = pair.split(",")[1].split("-")
        e1_range = range(int(e2[0]), int(e2[1]) + 1)
        e2_range = range(int(e1[0]), int(e1[1]) + 1)
        overlap = set(e1_range) & set(e2_range)
        if len(overlap) in [len(e1_range), len(e2_range)]:
            p1_total += 1
        # ---- Part 2 ----
        if len(overlap) > 0:
            p2_total += 1

print(f"Part 1: {p1_total}\nPart 2: {p2_total}")

File: src/d4.R
# Load lines of text file
data <- readLines("../data/d4.txt")

# Counters
p1_total <- 0
p2_total <- 0

for (line in data){
  e1 <- unlist(strsplit(unlist(strsplit(line, ','))[1], '-'))
  e2 <- unlist(strsplit(unlist(strsplit(line, ','))[2], '-'))
  z1 <- seq(e1[1], e1[2])
  z2 <- seq(e2[1], e2[2])
  if (length(intersect(z1, z2)) %in% c(length(z1), length(z2))){
    p1_total <- p1_total + 1
  }
  if (length(intersect(z1, z2)) > 0){
    p2_total <- p2_total + 1
  }
}
print(paste0('Part 1: ', p1_total, ', Part 2: ', p2_total))